text/0009-BundleTracker-integration.md

```diff
@@ -0,0 +1,184 @@
+- Start Date: 2022-05-25
+- RFC PR: (in a subsequent commit to the PR, fill me with the PR's URL)
+- CppMicroServices Issue: (fill me with the URL of the CppMicroServices issue that necessitated this RFC, otherwise leave blank)
+
+# Integration of BundleTracker into Core Framework
+
+## Summary
+
+This document describes the design and integration of the BundleTracker into the CppMicroServices core framework. The BundleTracker functionality is described in the [OSGI BundleTracker Specification](http://docs.osgi.org/specification/osgi.core/7.0.0/util.tracker.html#d0e52020), which is mapped to an API definition in C++ described in this document. This document discusses the need for adding BundleTracker to the core framework. Namely, there is existing compendium code that uses workarounds for lack of a BundleTracker in the core framework (e.g. bundle extension creation/destruction in DeclarativeServices), that would be made simpler and more robust with a proper BundleTracker implementation. Additionally, a chosen design and alternative implementations for BundleTracker are discussed.
+
+## Motivation
+
+Generally, it is useful to be able to react to state changes in bundles. The core framework offers a BundleListener to do this, but it is limited in its functionality on its own.
+
+The main issue with the BundleListener is that it only gives the user information on the bundle state changes _after_ the listener is initialized, not the full history of bundle states. Therefore, the user will not find out about any bundles that have been in a certain state since before the BundleListener was started. The BundleTracker, on the other hand, provides the full history of states for bundles (initial, transitory, and current).
+
+To discover this prior information without a BundleTracker, a user might loop through the bundles either before or after starting the BundleListener. Both methods have unexpected results in a multithreaded context, and solving the race conditions while avoiding deadlock is nontrivial.
+
+Currently, Declarative Services gets around this issue by starting a BundleListener and then calling tracker methods on each bundle in a controlled way (in the SCRActivator). However, this is a fragile workaround that is easy to implement incorrectly, leading to race conditions or deadlock. As such, it seems appropriate to implement a BundleTracker based on the OSGi specification to handle this problem inside of the core framework.
```

Comment:
The language here is amended to be more explicit. The DS implementation does try to get a record of the already active bundles by looping through the bundles already in the context before registering the listener. It is a flawed workaround though because of the possible race conditions described in the OSGi spec and the RFC, where events could get "double-counted" by both the loop and the listener if the state changes. The use of a BundleTracker there would fix that problem.

```
// Add bundle listener
  bundleListenerToken = context.AddBundleListener(
    std::bind(&SCRActivator::BundleChanged, this, std::placeholders::_1));
  // HACK: Workaround for lack of Bundle Tracker. Iterate over all bundles and call the tracker method manually
  for (const auto& bundle : context.GetBundles()) {
    if (bundle.GetState() == cppmicroservices::Bundle::State::STATE_ACTIVE) {
      cppmicroservices::BundleEvent evt(
        cppmicroservices::BundleEvent::BUNDLE_STARTED, bundle);
      BundleChanged(evt);
    }
  }
```